Improve Sieve of Eratosthenes article: expanded explanations, added code and comments for odd-only and optimized sieves, and corrected mathematical typos. (#1483)
cp-algorithms/cp-algorithms: src/algebra/sieve-of-eratosthenes.md

# Sieve of Eratosthenes

Sieve of Eratosthenes is an algorithm for finding all the prime numbers in a segment $[1, n]$ using $O(n \log \log n)$ operations.

The algorithm is very simple:
at the beginning we write down all numbers between 2 and $n$.
We mark all proper multiples of 2 (since 2 is the smallest prime number) as composite.

> A proper multiple of a number $n$, is a number greater than $n$ and divisible by $n$.

Reviewer: ```suggestion
A proper multiple of a number $m$, is a number greater than $n$ and divisible by $m$.
```

$n$ is already used.

### Sieving till root

Obviously, to find all the prime numbers until $n$, it will be enough just to perform the sifting only by the prime numbers, which do not exceed the root of $n$.

> To understand it more clearly please checkout RobJohn's [this answer](https://math.stackexchange.com/a/58808) to the question "Why in Sieve of Eratosthenes of $N$
> number you need to check and cross out numbers up to $\sqrt{N}$?" on [MathStackExchange](https://math.stackexchange.com).

Reviewer: ```suggestion
This is because the smallest prime factor $p$ of a composite number $m \leq n$ may not exceed $\sqrt m \leq \sqrt n$. Indeed, $p \cdot \frac{m}{p} = m$ and $p \leq \frac{m}{p}$, thus if $p$ was greater than $\sqrt m$, the product of $p$ and $\frac{m}{p}$ would be greater than $(\sqrt m)^2 = m$.
```

I think it's better to provide an inline explanation, rather than link to stackexchange, esp. given that the explanation is fairly short.

```cpp
int n;
vector<bool> is_prime(n+1, true);
is_prime[0] = is_prime[1] = false;
for (int i = 2; i * i <= n; i++) {
    if(!is_prime[i]) continue;

>     for (int j = i * i; j <= limit; j += i){

Reviewer: ```suggestion
    for (int j = i * i; j <= n; j += i){
```

`limit` is used here, but not defined anywhere. Generally, I don't think it should be different from $n$ in this particular part.

First, it will allow us to halve the needed memory. Second, it will reduce the number of operations performed by algorithm approximately in half.

```cpp
// returns a vector of all prime numbers in the interval [2, n]

> vector<int> primes_upto(int n){

Reviewer: Was this code tested on any particular problem? If no, could you please do it?

A benchmark ([link](https://gist.github.com/jakobkogler/e6359ea9ced24fe304f1a8af3c9bee0e)) shows, that using a `vector<bool>` is between 1.4x and 1.7x faster than using a `vector<char>`.

The same considerations also apply to `bitset`.

> It's also an efficient way of storing bits, similar to `vector<bool>`, so it takes only $\frac{N}{8}$ bytes of memory, but is a bit slower in accessing the elements means `vector<bool>` is a space-optimized specialization that typically uses 1 bit per element, unlike standard containers like `vector<char>` or `vector<bool>` (if it were a normal template), which use 1 byte per element.

Reviewer: ```suggestion
It's also an efficient way of storing bits, similar to `vector<bool>`, so it takes only $\frac{N}{8}$ bytes of memory, but is a bit slower in accessing the elements.
```
This paragraph is about `bitset`, not `vector<bool>`. The explanation in this particular paragraph seems out of place, moreover I believe the same information about `vector<bool>` is already conveyed above, on lines 159-160.

In the benchmark above `bitset` performs a bit worse than `vector<bool>`.
Another drawback from `bitset` is that you need to know the size at compile time.

To maintain the consistency and readability in this article we will keep using `vector<bool>`.

### Segmented Sieve

It follows from the optimization "sieving till root" that there is no need to keep the whole array `is_prime[1...n]` at all times.
For sieving it is enough to just keep the prime numbers until the root of $n$, i.e. `prime[1... sqrt(n)]`, split the complete range into blocks, and sieve each block separately.

> Let $s$ be a constant which determines the size of the block, then we have $\lceil \frac{n}{s} \rceil$ blocks altogether, and each block $k$ contains $s$ numbers in the interval $[ks, (k+1)s)$ for $k = 0, 1, \dots , \lfloor \frac{n}{s} \rfloor$.

Reviewer: The last block ends in $n$, rather than $(k+1)s$, right? I'm not sure if it needs to be mentioned separately though.

> We can work on blocks by turns, i.e. for every block $k$ we will go through all the prime numbers (from $1$ to $\sqrt n$) and remove mark their multiple as composite.

Reviewer: ```suggestion
We can work on blocks by turns, i.e. for every block $k$ we will go through all the prime numbers (from $1$ to $\sqrt n$) and mark their multiple as composite.
```

While working on the last block it should not be forgotten that the last needed number $n$ is not necessarily located at the end of the block.

As discussed previously, the typical implementation of the Sieve of Eratosthenes is limited by the speed how fast you can load data into the CPU caches.
By splitting the range of potential prime numbers $[1; n]$ into smaller blocks, we never have to keep multiple blocks in memory at the same time, and all operations are much more cache-friendlier.
As we are now no longer limited by the cache speeds, we can replace the `vector<bool>` with a `vector<char>`, and gain some additional performance as the processors can handle read and writes with bytes directly and don't need to rely on bit operations for extracting individual bits.
The benchmark ([link](https://gist.github.com/jakobkogler/e6359ea9ced24fe304f1a8af3c9bee0e)) shows, that using a `vector<char>` is about 3x faster in this situation than using a `vector<bool>`.

> A word of caution: those numbers might differ depending on architecture, compiler, and optimization levels.

Reviewer: Why do you think it's better to put it below the code, than where it was before?

> int count_primes(int n, int s = 1000){

Reviewer: Did you test the new version on any particular problem? If no, could you please do it?

    
    vector<int> primes;

>     vector<bool> is_prime(limit + 1, true);

Reviewer: > As we are now no longer limited by the cache speeds, we can replace the `vector<bool>` with a `vector<char>`

The usage of `vector<char>` rather than `vector<bool>` was intentional.

    // genetating primes in range [2, sqrt(n)]
    for(int i = 2; i <= limit; i++) {
        if(!is_prime[i]) continue;
        primes.push_back(i);
        for(int j = i * i; j <= limit; j += i){
            is_prime[j] = false;
        }
    }

    int total_primes = 0;
    for (int k = 0; k * s <= n; k++){
        // current block's range [left, right)
        int left = k * s, right = (k + 1) * s;

        // block[i] denotes whether the number (left + i) is prime.
        vector<bool> block(s, true);
        if (k == 0) block[0] = block[1] = false;

        for(int prime : primes) {
            // start is first multiple of 'prime' >= left.
            // see the note at the end of this code to understand the initialization of 'start'.
            int start = max(prime * prime, (left + prime - 1) / prime * prime);
            for(int j = start; j < right; j+=prime){
                block[j - left] = false;
            }
        }

        for(int i = 0; i < s && left + i <= n; i++){
            if(block[i]) total_primes++;
        }
    }
       
    return total_primes;
}
```

> #### Note on 'start'

Reviewer: I'm not sure it warrants a dedicated section. Also when your additions ends, it is followed by some general stuff that is not from this particular section, which might be confusing as to where we stop talking about `start` and start discussing the block sieve in general.

> As we know that we starting marking the multiple of a prime $p$ as composite starting from $p^2$. And the multiple of $p$ that we mark starting from $p^2$ are $p^2, p^2 + p, p^2 + 2p, \dots$ are of the form $p(p + i)$ for $i \geq 0$.

Reviewer: ```suggestion
We start marking the multiple of a prime $p$ as composite starting from $p^2$. The multiples of $p$ that we mark starting from $p^2$ are $p^2, p^2 + p, p^2 + 2p, \dots$, in other words the numbers of the form $p(p + i)$ for $i \geq 0$.
```

Now to start marking multiple of $p$ in range $[left, right)$ we first need to figure out the first multiple of $p \geq$ *left*.

> If *left* lies between two consecutive multiple of $p$ as $[p(p+i), p(p+i+1)]$ then we should starting marking numbers starting from $p(p+i+1)$ which is given by $\left\lceil \frac{left}{p} \right\rceil \cdot p $ that is `((left + p - 1) / p) * p`.

Reviewer: ```suggestion
If *left* lies between two consecutive multiple of $p$ as $[p(p+i), p(p+i+1)]$ then we should start marking numbers starting from $p(p+i+1)$ which is given by $\left\lceil \frac{left}{p} \right\rceil \cdot p $ that is `((left + p - 1) / p) * p`.
```

We pre-generate all prime numbers up to $\sqrt R$, and use those primes to mark all composite numbers in the segment $[L, R]$.

```cpp

> vector<bool> segmentedSieve(long long L, long long R) {

Reviewer: > As we are now no longer limited by the cache speeds, we can replace the `vector<bool>` with a `vector<char>`

The usage of `vector<char>` rather than `vector<bool>` was intentional.

> vector<bool> segmentedSieve(long long L, long long R) {

Reviewer: Did you test the new version on any particular problem? If no, could you please do it?

    // generate all primes up to sqrt(R)
    long long limit = sqrt(R);
    vector<long long> primes;
    vector<bool> is_prime(limit + 1, true);

    for(long long i = 2; i <= limit; ++i){
        if(!is_prime[i]) continue;
        primes.push_back(i);
        for(long long j = i * i; j <= limit; j += i){
           is_prime[j] = false;
        }
    }

    // primes_in_range[i] is true is L + i is prime where i belongs to [0, R-L].
    vector<bool> primes_in_range(R - L + 1, true);
    if(L == 1) primes_in_range[0] = false;

    for(long long prime : primes){
        long long start = max(prime * prime, (L + prime - 1) / prime * prime);
        for(long long j = start; j <= R; j+=prime){
            primes_in_range[j - L] = false;
        }
    }
        
    return primes_in_range;
}
```
Time complexity of this approach is $O((R - L + 1) \log \log (R) + \sqrt R \log \log \sqrt R)$.

It's also possible that we don't pre-generate all prime numbers:

```cpp

> vector<bool> segmentedSieveNoPreGen(long long L, long long R) {

Reviewer: Did you test the new version on any particular problem? If no, could you please do it?
